include/seqan3/alignment/pairwise/detail/concept.hpp

```diff
@@ -14,35 +14,78 @@
+ */
+//!\cond
+template <typename t>
+SEQAN3_CONCEPT sequence_pair = requires ()
+{
+    requires tuple_like<t>;
+    requires std::tuple_size_v<t> == 2;
+    requires std::ranges::forward_range<std::tuple_element_t<0, t>>;
+    requires std::ranges::forward_range<std::tuple_element_t<1, t>>;
+    requires semialphabet<std::ranges::range_value_t<std::tuple_element_t<0, t>>>;
+    requires semialphabet<std::ranges::range_value_t<std::tuple_element_t<1, t>>>;
+};
+//!\endcond
+
+/*!\interface seqan3::detail::sequence_pair_range <>
+ * \brief A helper concept to check if a type is a range over seqan3::detail::sequence_pair's.
+ * \ingroup pairwise_alignment
+ *
+ * \tparam t The type to check.
+ *
+ * \details
+ *
+ * This concept checks if the given type models a std::ranges::forward_range over and that the value type of the
+ * range models seqan3::detail::sequence_pair.
+ */
+//!\cond
+template <typename t>
+SEQAN3_CONCEPT sequence_pair_range = std::ranges::forward_range<t> && sequence_pair<std::ranges::range_value_t<t>>;
+//!\endcond
+
+/*!\interface seqan3::detail::indexed_sequence_pair_range <>
+ * \brief A helper concept to check the input of the range based alignment algorithm interface.
  * \ingroup pairwise_alignment
  *
  * \tparam t The type to check.
  *
  * \details
  *
  * This concept checks if the given type models a std::ranges::forward_range over indexed sequence pairs that are
- * passed to the alignment algorithms. An indexed sequence pair is a pair of sequences that shall be aligned. In
- * addition, the pair is annotated with an index so that the caller can infer the aligned sequences from the returned
- * seqan3::alignment_result. The layout of this indexed sequence type looks as follows:
- * * the first type of the pair refers to the sequence pair type, and
+ * passed to the alignment algorithms. An indexed sequence pair consists of a seqan3::detail::sequence_pair
+ * that shall be aligned and an index that is used to identify the aligned sequence pair.
+ * The caller can then infer the aligned sequences from the returned seqan3::alignment_result.
+ * The layout of this indexed sequence type looks as follows:
+ * * the first type of the pair must model seqan3::detail::sequence_pair, and
  * * the second type of the pair refers to the respective index type.
  */
 //!\cond
 template <typename t>
-SEQAN3_CONCEPT indexed_sequence_pair_range = requires (std::ranges::range_value_t<t> value)
+SEQAN3_CONCEPT indexed_sequence_pair_range = std::ranges::forward_range<t> &&
+                                             requires (std::ranges::range_value_t<t> value)
 {
-    requires std::ranges::forward_range<t>;
     requires tuple_like<decltype(value)>;
     requires std::tuple_size_v<decltype(value)> == 2;
-    requires tuple_like<std::tuple_element_t<0, decltype(value)>>;
-    requires std::tuple_size_v<std::tuple_element_t<0, decltype(value)>> == 2;
+    requires sequence_pair<std::tuple_element_t<0, decltype(value)>>;
```

Comment:
Idea: Should we require some numeric/integral type or at least EqualityComparable for the index? Otherwise users are able to set weird types, which may cause side effects at some point.

Reply:
The thing is I am not yet completey sure what the requested index type should be. Is it always just a numeric value? It certainly will be at some point an pair over numeric values. Can it be something more complex? As a matter of fact, we also don't care. We just propagate the index as is to the alignment result. So we don't do any operations on that. The only thing we can require is copyability. That is something I can indeed test.

Reply:
Okay, that's a good solution.